confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+        self._schema = Schema(_schema_to_str(proto_file),
+                              schema_type='PROTOBUF')
+
+    @staticmethod
+    def _encode_index(buf, index):
+        """
+        Writes the index to buf formatted as unsigned varints.
+
+        Args:
+            buf (BytesIO): buffer to write index to
+            index [index]: Message index
+
+        Returns:
+            [bytes]: Serialized Message index
+
+        """
+        for value in index:
+            while (value & 0x7f) != 0:
+                buf.write(_bytes((value & 0x7f) | 0x80))
+                value >>= 7
+            buf.write(_bytes(value))
+
+    def resolve_dependencies(self, ctx, file_desc):
+        """
+        Resolves and optionally registers schema references recursively.
+
+        Args:
+            ctx (SerializationContext): Serialization context.
+
+            file_desc (FileDescriptor): file descriptor to traverse.
+
+        """
+        schema_refs = []
+        for dep in file_desc.dependencies:
+            dep_refs = self.resolve_dependencies(ctx, dep)
+            subject = self._ref_reference_subject_func(ctx, dep)
+            schema = Schema(_schema_to_str(dep),
+                            references=dep_refs,
+                            schema_type='PROTOBUF')
+            if self._auto_register:
+                self._registry.register_schema(subject, schema)
+
+            reference = self._registry.lookup_schema(subject, schema)
+            # schema_refs are per file descriptor
+            schema_refs.append(SchemaReference(dep.name,
+                                               subject,
+                                               reference.version))
+        return schema_refs
+
+    def __call__(self, msg, ctx):
+        """
+        Serializes a Protobuf Message to the Confluent Schema Registry
+        Protobuf binary format.
+
+        Args:
+            msg (Message): Protobuf message instance.
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted Protobuf bytes
+
+        """
+        if msg is None:
+            return None
+
+        subject = self._subject_name_func(ctx, msg.DESCRIPTOR.full_name)
+
+        if subject not in self._known_subjects:
+            self._schema.references = self.resolve_dependencies(ctx, msg.DESCRIPTOR.file)
+
+        self._schema_id = self._registry.register_schema(subject, self._schema)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record index to the buffer
+            self._encode_index(fo, self.msg_index)
```

Comment:
this `msg_index` relates to the message descriptor passed into the constructor, however your code here will happily work if `msg` is any protobuf type. if a different type is serialized than the descriptor passed to the constructor, this will be incorrect.

Reply:
you haven't implemented the special case where the index array is [0]. This is encoded as 0, not 1, 0. see the C# code for e.g. impl: https://github.com/confluentinc/confluent-kafka-dotnet/blob/1.4.x/src/Confluent.SchemaRegistry.Serdes.Protobuf/ProtobufSerializer.cs#L147

Reply:
I actually handle this while generating the message index 

```
def _create_msg_index(msg_desc):
...
    # The root element at the 0 position does not need a length prefix.
    if len(msg_idx) == 1 and msg_idx[0] == 0:
        return [0]

    msg_idx.appendleft(len(msg_idx))
    return list(msg_idx)
```

I use a double ended queue

Reply:
oh right. my (minor) comment here then is if the steps were more clearly spelled out, i.e. 1. create the index array 2. modify for the optimization, i would have seen this easier. the more specific the function the better.

Reply:
I'd argue the "special case" here is really a characteristic of the index itself not the act of encoding the index which would make this the right place for it.

Reply:
yeah, agree with you. sorry, i wasn't giving this enough focus.

Reply:
Tripped up Magnus as well, renamed to `_encode_uvarints` in hopes it will more clearly convey the purpose of this method

Reply:
The optimization is a wire coding optimization, afaik, and should thus be done in the encoding function rather than this here.